modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadWriteTransactionImpl.java

```diff
@@ -159,18 +162,54 @@ public CompletableFuture<Void> finish(boolean commit, @Nullable HybridTimestamp
             return finishFuture;
         }
 
+        return finishInternal(commit, executionTimestamp, full, true);
+    }
+
+    private CompletableFuture<Void> finishInternal(
+            boolean commit,
+            @Nullable HybridTimestamp executionTimestamp,
+            boolean full,
+            boolean isComplete
+    ) {
         enlistPartitionLock.writeLock().lock();
 
         try {
             if (finishFuture == null) {
+                if (killed) {
+                    if (isComplete) {
+                        finishFuture = nullCompletedFuture();
+
+                        return failedFuture(new TransactionException(
```

Comment:
Why not set `failedFuture` to `finishFuture`?

Reply:
Because `finishFuture` is set as `nullCompletedFuture()`.
This matches with the general logic: an exception is thrown only in the first complete operation call (commit or rollback), the other calls after would be completed silently.